br/pkg/lightning/backend/external/bench_test.go

```diff
@@ -0,0 +1,211 @@
+	return s
+}
+
+type kvSource interface {
+	next() (key, value []byte, handle kv.Handle)
+	outputSize() int
+}
+
+type ascendingKeySource struct {
+	keySize, valueSize int
+	count              int
+	kvChan             chan [2][]byte
+	curKey             []byte
+	totalSize          int
+}
+
+func newAscendingKeySource(keySize, valueSize, count int) *ascendingKeySource {
+	s := &ascendingKeySource{keySize: keySize, valueSize: valueSize, count: count}
+	s.curKey = make([]byte, keySize)
+	s.kvChan = make(chan [2][]byte, 100)
+	s.run()
+	return s
+}
+
+func (s *ascendingKeySource) run() {
+	go func() {
+		defer close(s.kvChan)
+		for i := 0; i < s.count; i++ {
+			for j := len(s.curKey) - 1; j >= 0; j-- {
+				s.curKey[j]++
+				if s.curKey[j] != 0 {
+					break
+				}
+			}
+			key := make([]byte, s.keySize)
+			copy(key, s.curKey)
+			value := make([]byte, s.valueSize)
+			s.kvChan <- [2][]byte{key, value}
+			s.totalSize += len(key) + len(value)
+		}
+	}()
+}
+
+func (s *ascendingKeySource) next() (key, value []byte, handle kv.Handle) {
+	pair, ok := <-s.kvChan
+	if !ok {
+		return nil, nil, nil
+	}
+	return pair[0], pair[1], nil
+}
+
+func (s *ascendingKeySource) outputSize() int {
+	return s.totalSize
+}
+
+type testSuite struct {
+	t                  *testing.T
+	store              storage.ExternalStorage
+	source             kvSource
+	memoryLimit        int
+	beforeCreateWriter func()
+	beforeWriterClose  func()
+	afterWriterClose   func()
+}
+
+func writePlainFile(s *testSuite) {
+	ctx := context.Background()
+	buf := make([]byte, s.memoryLimit)
+	offset := 0
+	flush := func(w storage.ExternalFileWriter) {
+		n, err := w.Write(ctx, buf[:offset])
+		require.NoError(s.t, err)
+		require.Equal(s.t, offset, n)
+		offset = 0
+	}
+
+	if s.beforeCreateWriter != nil {
+		s.beforeCreateWriter()
+	}
+	writer, err := s.store.Create(ctx, "test/plain_file", nil)
+	require.NoError(s.t, err)
+	key, val, _ := s.source.next()
+	for key != nil {
+		if offset+len(key)+len(val) > len(buf) {
+			flush(writer)
+		}
+		offset += copy(buf[offset:], key)
+		offset += copy(buf[offset:], val)
+		key, val, _ = s.source.next()
+	}
+	flush(writer)
+	if s.beforeWriterClose != nil {
+		s.beforeWriterClose()
+	}
+	require.NoError(s.t, writer.Close(ctx))
+	if s.afterWriterClose != nil {
+		s.afterWriterClose()
+	}
+}
+
+func writeExternalFile(s *testSuite) {
+	ctx := context.Background()
+	builder := NewWriterBuilder().
+		SetMemorySizeLimit(uint64(s.memoryLimit))
+
+	if s.beforeCreateWriter != nil {
+		s.beforeCreateWriter()
+	}
+	writer := builder.Build(s.store, "test/external", "writerID")
+	key, val, h := s.source.next()
+	for key != nil {
+		err := writer.WriteRow(ctx, key, val, h)
+		require.NoError(s.t, err)
+		key, val, h = s.source.next()
+	}
+	if s.beforeWriterClose != nil {
+		s.beforeWriterClose()
+	}
+	require.NoError(s.t, writer.Close(ctx))
+	if s.afterWriterClose != nil {
+		s.afterWriterClose()
+	}
+}
+
+func TestCompare(t *testing.T) {
```

Comment:
Use `BenchmarkXXX` instead?

Reply:
I have used t.Skip() in line 33. Another reason is that I want to do more calculation in the tests, like compute the throughput ratio of external write / direct writing, and output profile files only for part of code. Benchmark will not help much about these purpose.